### PR TITLE
Fix and cleanup container metrics

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR:
1) Fix a potential panic in `ContainerStats`;
2) Fix a incorrect check `r.GetFilter == nil`, `GetFilter` is a function;
3) Refactor the code a little bit, so that we only access container store once to avoid the race;
4) Simplify the code a little bit.

Signed-off-by: Lantao Liu <lantaol@google.com>

### CODE REVIEW

**Comment by Random-Liu** [comment]
@abhinandanpb

File: pkg/server/container_stats.go [hunk 1]
```diff
@@ -32,19 +32,22 @@ func (c *criContainerdService) ContainerStats(ctx context.Context, in *runtime.C
...
+	if len(resp.Metrics) != 1 {
+		return nil, fmt.Errorf("unexpected metrics response: %+v", resp.Metrics)
 	}
 
-	var cs runtime.ContainerStats
-	if err := c.getContainerMetrics(containerID, resp.Metrics[0], &cs); err != nil {
-		return nil, fmt.Errorf("failed to decode container metrics: %v", err)
+	cs, err := c.getContainerMetrics(cntr.Metadata, resp.Metrics[0])
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate container metrics: %v", err)
```
Comment (abhinandanpb) [inline pkg/server/container_stats.go]: I would stick with decode. We are not technically generating :)

File: pkg/server/container_stats.go [hunk 2]
```diff
@@ -32,19 +32,22 @@ func (c *criContainerdService) ContainerStats(ctx context.Context, in *runtime.C
...
+		return nil, fmt.Errorf("failed to find container: %v", err)
 	}
 	request := &tasks.MetricsRequest{Filters: []string{"id==" + containerID}}
 	resp, err := c.taskService.Metrics(ctx, request)
 	if err != nil {
-		return nil, fmt.Errorf("failed to fetch metrics for tasks: %v", err)
+		return nil, fmt.Errorf("failed to fetch metrics for task: %v", err)
+	}
+	if len(resp.Metrics) != 1 {
+		return nil, fmt.Errorf("unexpected metrics response: %+v", resp.Metrics)
```
Comment (abhinandanpb) [inline pkg/server/container_stats.go]: Thanks for taking care of this

File: pkg/server/container_stats.go [hunk 3]
```diff
@@ -32,19 +32,22 @@ func (c *criContainerdService) ContainerStats(ctx context.Context, in *runtime.C
...
 	}
 
-	var cs runtime.ContainerStats
-	if err := c.getContainerMetrics(containerID, resp.Metrics[0], &cs); err != nil {
-		return nil, fmt.Errorf("failed to decode container metrics: %v", err)
+	cs, err := c.getContainerMetrics(cntr.Metadata, resp.Metrics[0])
```
Comment (abhinandanpb) [inline pkg/server/container_stats.go]: Thats fair :)

File: pkg/server/container_stats_list.go [hunk 4]
```diff
@@ -126,36 +112,33 @@ func (c *criContainerdService) getContainerMetrics(
...
+	if r.GetFilter() == nil {
 		return req, nil, nil
 	}
-
-	candidateContainers := make(map[string]bool)
-	for _, c := range c.containerStore.List() {
-		if r.Filter.GetId() != "" && c.ID != r.Filter.GetId() {
+	var containers []containerstore.Container
+	for _, cntr := range c.containerStore.List() {
+		if r.Filter.GetId() != "" && cntr.ID != r.Filter.GetId() {
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats_list.go]: Will do.

File: pkg/server/container_stats.go [hunk 5]
```diff
@@ -32,19 +32,22 @@ func (c *criContainerdService) ContainerStats(ctx context.Context, in *runtime.C
 		return nil, fmt.Errorf("invalid container stats request")
 	}
 	containerID := in.GetContainerId()
-	_, err := c.containerStore.Get(containerID)
+	cntr, err := c.containerStore.Get(containerID)
 	if err != nil {
-		return nil, fmt.Errorf("failed to find container %q: %v", containerID, err)
+		return nil, fmt.Errorf("failed to find container: %v", err)
 	}
 	request := &tasks.MetricsRequest{Filters: []string{"id==" + containerID}}
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats.go]: Will do.

File: pkg/server/container_stats.go [hunk 6]
```diff
@@ -32,19 +32,22 @@ func (c *criContainerdService) ContainerStats(ctx context.Context, in *runtime.C
 		return nil, fmt.Errorf("invalid container stats request")
 	}
 	containerID := in.GetContainerId()
-	_, err := c.containerStore.Get(containerID)
+	cntr, err := c.containerStore.Get(containerID)
 	if err != nil {
-		return nil, fmt.Errorf("failed to find container %q: %v", containerID, err)
+		return nil, fmt.Errorf("failed to find container: %v", err)
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats.go]: The assumption that the caller always know what id they are using. We only include information which caller doesn't know in the error message.

File: pkg/server/container_stats.go [hunk 7]
```diff
@@ -32,19 +32,22 @@ func (c *criContainerdService) ContainerStats(ctx context.Context, in *runtime.C
 		return nil, fmt.Errorf("invalid container stats request")
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats.go]: Will do.

File: pkg/server/container_stats_list.go [hunk 8]
```diff
@@ -23,25 +23,26 @@ import (
...
+	containerstore "github.com/kubernetes-incubator/cri-containerd/pkg/store/container"
 )
 
 // ListContainerStats returns stats of all running containers.
 func (c *criContainerdService) ListContainerStats(
 	ctx context.Context,
 	in *runtime.ListContainerStatsRequest,
 ) (*runtime.ListContainerStatsResponse, error) {
-	request, candidateContainers, err := c.buildTaskMetricsRequest(in)
+	request, containers, err := c.buildTaskMetricsRequest(in)
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats_list.go]: Haven't looked at your PR yet. Will do after the v1.0.0-alpha.0 release, and we could discuss there. :)

File: pkg/server/container_stats.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats.go]: Done

File: pkg/server/container_stats.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats.go]: Done

File: pkg/server/container_stats_list.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_stats_list.go]: Done.

File: pkg/server/container_stats.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/server/container_stats.go]: Will do.

**Comment by abhi** [comment]
LGTM

**Pull request merged**